content/name-based-xml.md

```diff
@@ -0,0 +1,421 @@
+    def literal[+E <: Element](builder: Builder[E]) = builder.element
+  }
+}
+```
+
+The ability of custom implementation for XML literals enables a lot of possibilities, including React-like virtual DOM frameworks, static type checked XML schema, and reactive XHTML/FXML templating.
+
+### Comparison Examples
+
+XML literals in Scala 2.13 are symbol based, which means the type of an XML literal is always `scala.xml.Elem`. The code `<div title="my-title">line1<br/>line2</div>` will be parsed to an AST equivalent to the following code:
+
+``` scala
+{
+  var $md: MetaData = Null;
+  $md = new UnprefixedAttribute("title", new Text("my-title"), $md);
+  new Elem(null, "div", $md, TopScope, false, ({
+    val $buf = new NodeBuffer();
+    $buf.$amp$plus(new Text("line1"));
+    $buf.$amp$plus({
+      {
+        new Elem(null, "br", Null, TopScope, true)
+      }
+    });
+    $buf.$amp$plus(new Text("line2"));
+    $buf
+  }: _*))
+}
+```
+
+Note that `MetaData`, `UnprefixedAttribute`, `Text`, `Elem` and `TopScope` are types defined in `scala.xml`. Library authors cannot create their custom representation of XML literals.
+
+### Counter-Examples
+
+With the help of this proposal, a library author can create custom XML based DSL.
+
+``` scala
+<div><label v-if={math.random() < 0.5}>You might see me</label></div>
+```
+
+It will be translated to:
+
+``` scala
+xml.literal(
+  xml.elements.div(
+    xml.elements.label(
+      xml.attributes.`v-if`(xml.interpolation(math.random() < 0.5)),
+      xml.texts.`You might see me`
+    )
+  )
+)
+```
+
+The translated `v-if` method can be supported by creating an `implicit class` for `xml.attributes` easily. However, this proposal is not intended to be used for creating a DSL as an alternative to regular Scala code. Instead, you should just use ordinary `if` expression.
+
+``` scala
+<div>{
+  if (math.random() < 0.5) {
+    <label>You might see me</label>
+  } else {
+    <!--condition not met-->
+  }
+}</div>
+```
+
+and it will be translated to:
+
+``` scala
+xml.literal(
+  xml.elements.div(
+    xml.interpolation {
+      if (math.random() < 0.5) {
+        xml.literal(
+          xml.elements.label(
+            xml.texts.`You might see me`
+          )
+        )
+      } else {
+        xml.literal(
+          xml.comment("condition not met")
+        )
+      }
+    }
+  )
+)
+```
+
+Name based XML literal is best for modeling existing XML format within Scala source files. For other cases, just use ordinary Scala control flows for dynamic conditions, and use ordinary Scala case class for data structures not defined in the existing XML format.
+
+## Design
+
+### Goals
+
+* Keeping source-level backward compatibility to existing symbol-based XML literals in most use cases of `scala-xml`.
+* Allowing schema-aware XML literals, i.e. static type varying according to tag names, similar to the current TypeScript and [Binding.scala](https://github.com/ThoughtWorksInc/Binding.scala) behavior.
```

Comment:
I am not a TypeScript expert. Do you know more about how TypeScript type-check JSX? What does "black box" actually mean? Do they translate XML literals to JS before type-checking or after?

Reply:
I consider myself an expert TypeScript *user*, so my knowledge about compiler internals is partial, but the main issue with JSX in TS is that the result type of any JSX expression is `JSX.Element`, which is monomorphic.

This means you can't check things like:

```tsx
<MyList>
  <ListItem /> // ok
  <WrongComponent /> // can't get an error here
</MyList>
```

because all of those tags will just have `JSX.Element` type.

I suspect TypeScript type-checks JSX before emitting any JS, since it even has the option of preserving JSX in the emitted code (see the `jsx: "preserve"` compiler option), but you still get type-checking in that case.

TypeScript allows to swap out the implementation it emits JSX to, e.g. the default is `React.createElement`, but you can choose a different one when using alternative libraries that make use of JSX (like preact, Vue, etc). This does not affect type checking, since they all just must return `JSX.Element` anyway.

There are a bunch more details to how TypeScript works (e.g. how it manages function vs classes for component definitions, how it distinguishes between native and user-defined elements, and so on) but I think the main take is that it's ultimately quite a bit different to what it's being proposed here.

The very core of the proposal (being able to get different types for different XML tags) is precisely what TypeScript is lacking.